### PR TITLE
Equipment list to show Status

### LINKED ISSUE #3140: Asset list view part 2
Part of epic #2896 
## Is your feature request related to a problem? Please describe 👀
[Figma designs](https://www.figma.com/file/rIK2zVFUVhtbuXo0RBzOzw/0033-2305_omSupply-ColdChain-UI?type=design&node-id=2653-16424&mode=design&t=8KRUuAo7s1V0r0RM-0)

Follow up work from #3137

## Describe the solution you'd like 🎁

* [Adam's comment](https://github.com/msupply-foundation/open-msupply/issues/2973#issuecomment-1960717319) on colours of the status indication
* Status column
* ~Manufacturer & Model~
* ~Batch deletions~
* ~Scan? Perhaps a new issue for this one?~

### LINKED ISSUE #3204: Add enum of reasons to asset log
Hard coded list of reasons mapped to a particular status:

See spreadsheet for details:

https://docs.google.com/spreadsheets/d/1l1x38KT5cdt4Zd5piMWp3FuBuiSUv4faMHP8utDJDqU/edit#gid=2113323501

### Moneyworks Jobcode 🧰

OMSCC:AST

### PR DESCRIPTION
**Pull request by mark-prins**
Fixes #3140 

# 👩🏻‍💻 What does this PR do? 
 
Adds a loader to get the current status of an asset and shows the status in the equipment list like this:

image

# 🧪 How has/should this change been tested? 

For this, I've manually created asset logs which are associated to my equipment. Alternatively, you could wait until the next PR if you want to have the UI for adding, or use the API directly.

## 📃 Documentation

_Part of the CCEI feature_

### CODE REVIEW

File: server/graphql/core/src/loader/asset_status.rs [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        let repo = AssetLogRepository::new(&connection);
+
+        let filter = AssetLogFilter::new().asset_id(EqualFilter::equal_any(ids.to_owned()));
+
+        let sort = AssetLogSort {
+            key: AssetLogSortField::LogDatetime,
+            desc: Some(false),
+        };
+
+        let result = repo
```
Comment (mark-prins) [inline server/graphql/core/src/loader/asset_status.rs:30]: perhaps not the most efficient way to build the hash map.

Comment (jmbrunskill) [reply server/graphql/core/src/loader/asset_status.rs:30]: I'd be tempted to create a view like `changelog_deduped` or write a raw SQL query in diesel something like this...
```
CREATE VIEW latest_asset_log AS
  SELECT al.id,
    al.asset_id,
    al.user_id,
    al.status,
    al.comment,
    al.type,
    al.reason,
    al.log_datetime
  FROM (
    SELECT asset_id, MAX(log_datetime) AS latest_log_datetime
    FROM asset_log
    GROUP BY asset_id
  ) grouped
  INNER JOIN asset_log al
    ON al.asset_id = grouped.asset_id AND al.log_datetime = grouped.latest_log_datetime;

CREATE INDEX ix_asset_log_asset_id_log_datetime ON asset_log (asset_id, log_datetime);
```

Or just run a query for each ID maybe? Could be a lot but bounded to how many we display in a page (e.g. 20?)

I think reading all each asset's log entry just to get that latest could be a potential performance issue in future if there's 1000s of status logs...

I'm also not 100% if we're guaranteed to get the right result with this implementation. Have you found any docs to suggest that `collect()` will always take the first record for each id as the hashmap entry? I've had a look but can't find anything definitive...

File: server/repository/src/migrations/v1_08_00/assets/asset_log.rs [hunk 2]
```diff
@@ -8,8 +8,8 @@ pub(crate) fn migrate(connection: &StorageConnection) -> anyhow::Result<()> {
         CREATE TABLE asset_log (
             id TEXT NOT NULL PRIMARY KEY,
             asset_id TEXT NOT NULL REFERENCES asset(id),
-            user_id TEXT REFERENCES user_account(id),
-            status TEXT, 
+            user_id NOT NULL TEXT REFERENCES user_account(id),
+            status TEXT,
```
Comment (jmbrunskill) [inline server/repository/src/migrations/v1_08_00/assets/asset_log.rs:12]: Should we make this not null too?
Do we want to enter a lot with no status?

**Review by jmbrunskill — COMMENTED** [review]
Should have been a comment on an earlier change but I wonder if we want to make more fields required? Generally simplifies code if we don't have Option<> and Nullable flying around...

**Review by fergie-nz — APPROVED** [review]
Looks great.

I don't quite understand how the hashmap returns just the most recent log, but is working great!

File: server/repository/src/migrations/v1_08_00/assets/asset_log.rs [hunk 3]
```diff
@@ -8,8 +8,8 @@ pub(crate) fn migrate(connection: &StorageConnection) -> anyhow::Result<()> {
         CREATE TABLE asset_log (
             id TEXT NOT NULL PRIMARY KEY,
             asset_id TEXT NOT NULL REFERENCES asset(id),
-            user_id TEXT REFERENCES user_account(id),
-            status TEXT, 
+            user_id NOT NULL TEXT REFERENCES user_account(id),
```
[the remark this answers is not in the document]
Comment (jmbrunskill) [reply server/repository/src/migrations/v1_08_00/assets/asset_log.rs:11]: Yup that's a bug I think. Probably works in Postgres but SQLite doesn't handle it.

**Comment by jmbrunskill** [comment]
> I don't quite understand how the hashmap returns just the most recent log, but is working great!
I'm pretty sure the collect() function loops through every `vec` entry and tries to insert the value into the `HashMap` this effectively creates a last entry wins effect. I can't imagine this behaviour changing but I don't like it, both from the point of view of relying on this ordering and the implementation of collect() for HashMap, and because might be slow if there's a lot of logs to trawel through...

**Comment by mark-prins** [comment]
awesome! thanks for the feedback. have implemented the view and I'm much happier with that 🙏

**Pull request merged**